gobblin-yarn/src/main/java/gobblin/yarn/YarnService.java

```diff
@@ -0,0 +1,457 @@
+
+  private static final Splitter SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
+
+  private final ConcurrentMap<ContainerId, Container> containerMap = Maps.newConcurrentMap();
+  private final ConcurrentMap<ContainerId, ParticipantId> containerToParticipantMap = Maps.newConcurrentMap();
+
+  private final ApplicationId applicationId;
+  private final String applicationName;
+
+  private final Config config;
+
+  private final Configuration yarnConfiguration;
+  private final AMRMClientAsync<AMRMClient.ContainerRequest> amrmClientAsync;
+  private final NMClientAsync nmClientAsync;
+  private final ExecutorService containerLaunchExecutor;
+  private final FileSystem fs;
+
+  private final EventBus eventBus;
+
+  private final int initialContainers;
+  private final int requestedContainerMemoryMbs;
+  private final int requestedContainerCores;
+
+  private final String containerJvmArgs;
+
+  // Security tokens for accessing HDFS
+  private final ByteBuffer tokens;
+
+  private final Closer closer = Closer.create();
+
+  private final Object allContainersStopped = new Object();
+
+  private volatile Optional<Resource> maxResourceCapacity =Optional.absent();
+
+  public YarnService(Config config, String applicationName, ApplicationId applicationId, FileSystem fs,
+      EventBus eventBus, String containerJvmArgs) throws Exception {
+    this.applicationName = applicationName;
+    this.applicationId = applicationId;
+
+    this.config = config;
+
+    this.yarnConfiguration = new YarnConfiguration();
+    this.amrmClientAsync = closer.register(
+        AMRMClientAsync.createAMRMClientAsync(1000, new AMRMClientCallbackHandler()));
+    this.amrmClientAsync.init(this.yarnConfiguration);
+    this.nmClientAsync = closer.register(NMClientAsync.createNMClientAsync(new NMClientCallbackHandler()));
+    this.nmClientAsync.init(this.yarnConfiguration);
+
+    this.containerLaunchExecutor = Executors.newFixedThreadPool(10,
+        ExecutorsUtils.newThreadFactory(Optional.of(LOGGER), Optional.of("ContainerLaunchExecutor")));
+
+    this.fs = fs;
+
+    this.eventBus = eventBus;
+
+    this.initialContainers = config.getInt(GobblinYarnConfigurationKeys.INITIAL_CONTAINERS_KEY);
+    this.requestedContainerMemoryMbs = config.getInt(GobblinYarnConfigurationKeys.CONTAINER_MEMORY_MBS_KEY);
+    this.requestedContainerCores = config.getInt(GobblinYarnConfigurationKeys.CONTAINER_CORES_KEY);
+
+    this.containerJvmArgs = containerJvmArgs;
+
+    this.tokens = getSecurityTokens();
+
+    // Register itself with the EventBus for container-related requests
+    this.eventBus.register(this);
+  }
+
+  @SuppressWarnings("unused")
+  @Subscribe
+  public void handleNewContainerRequest(NewContainerRequest newContainerRequest) {
+    int newContainersRequested = newContainerRequest.getNewContainersRequested();
+    if (newContainersRequested <= 0) {
+      LOGGER.error("Invalid number of new containers requested: " + newContainersRequested);
```

Comment:
Change error message to "Number of containers requested me be positive and non-zero..."